content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,520 @@
+
+For example, lets say we have pretty dashboard, full of beautiful graphs that display meaningful data. Let's make it a line chart, that displays multiple lines, and we want to have the ability to toggle certain lines on and off.
+
+The data could look something like this:
+
+```js
+const veryImportantData = [
+  {name: 'Impressions', visible: true, data: [...]},
+  {name: 'Clicks', visible: false, data: [...]},
+  {name: 'PageViews', visible: true, data: [...]}
+]
+```
+
+We have already built a function using our newfound knowledge of `map` to toggle the `visible` state on and off for a line, but now we need to grab only the lines that have `visible: true`, and send them in to our chart for presentation. This is where `filter` comes in handy!
+
+```js
+function isLineVisible(line) {
+  return line.visible;
+}
+
+const dataToPresent = veryImportantData.filter(isLineVisible);
+```
+
+Or, we could do it inline with an anonymous function on one line!
+
+```js
+const dataToPresent = veryImportantData.filter(line => line.visible);
+```
+
+Which gives us a new array with only the lines that had `visible: true`. The rest have been filtered out. Pretty great, right?
+
+Compare this to an approach using `for` loops:
+
+```js
+const dataToPresent = [];
+
+for (let i = 0; i < veryImportantData.length; i++) {
+  if (veryImportantData[i].visible) {
+    dataToPresent.push(veryImportantData[i]);
+  }
+}
+```
+
+Once again, the higher order function version is easier to grasp, and we don't need to care about how the loop is working, or how the array is created. It's all abstracted away from us, focusing on the important part - filtering logic.
+
+### Array.reduce
+
+`Array.reduce`, while very useful, is a bit harder to grasp than `Array.map` and `Array.filter`, it is very versatile, and can be used to build new objects, numbers, strings, or arrays.
+
+It takes a reducer function as argument, that is run on every value. However, this reducer function has a slight difference with the former two functions we have talked about. It takes an `accumulator` as the first argument, and the element it's currently operating on. The `reduce` function also has a second argument, `initialValue`, which can be sent in to set the `accumulator` to have a value the first iteration.
+
+Lets look at an example. Lets take our shopping cart again. Say you want to present the total price of the shopping cart. This is a handy use for reduce!
+
+```js
+const toalPrice = products
+  .map(addTotalPriceToProduct)
+  .reduce((total, product) => total + product.totalPrice, 0);
+
+totalPrice; // 1275
+```
+
+So, first we run a map function on the products, adding totalPrice to each product. Secondly, we run the reduce function. Take a close look at the chaining. We can chain all these higher order functions on the `Array` object, as long as the return a new array.
```

Comment:
I'm not sure I agree, when we create a new array, it is an instance of `Array`, not `Array.prototype`. Using it for the functions makes sense since that's where they come from, but I think it might be confusing to say `Array.prototype` object/instance.

> The Array.prototype property represents the prototype for the Array constructor.